File: python/mirage/commands/builtin/generic/cmp.py
from collections.abc import Awaitable, Callable

from mirage.io.types import ByteSource, IOResult
from mirage.types import PathSpec


async def cmp_cmd(
    paths: list[PathSpec],
    *,
    read_bytes: Callable[..., Awaitable[bytes]],
    accessor: object = None,
    silent: bool = False,
    verbose: bool = False,
    limit: int | None = None,
    print_bytes: bool = False,
    skip: int | None = None,
) -> tuple[ByteSource | None, IOResult]:
    if len(paths) < 2:
        raise ValueError("cmp: requires two paths")
    p0, p1 = paths[0], paths[1]
    data1 = await read_bytes(accessor, p0)
    data2 = await read_bytes(accessor, p1)
    if skip is not None:
        data1 = data1[skip:]
        data2 = data2[skip:]
    if limit is not None:
        data1 = data1[:limit]
        data2 = data2[:limit]
    if data1 == data2:
        return None, IOResult()
    if silent:
        return None, IOResult(exit_code=1)
    if verbose:
        out_lines: list[str] = []
        for idx in range(min(len(data1), len(data2))):
            if data1[idx] != data2[idx]:
                out_lines.append(
                    f"{idx + 1} {oct(data1[idx])} {oct(data2[idx])}")
        return "\n".join(out_lines).encode(), IOResult(exit_code=1)
    for idx in range(min(len(data1), len(data2))):
        if data1[idx] != data2[idx]:
            line = 1 + data1[:idx].count(ord(b"\n"))
            msg = (f"{p0.original} {p1.original}"
                   f" differ: char {idx + 1}, line {line}")
            if print_bytes:
                msg += (f" is {oct(data1[idx])} {chr(data1[idx])}"
                        f" {oct(data2[idx])} {chr(data2[idx])}")
            return msg.encode(), IOResult(exit_code=1)
    shorter = p0.original if len(data1) < len(data2) else p1.original
    msg = f"cmp: EOF on {shorter}"
    return msg.encode(), IOResult(exit_code=1)


__all__ = ["cmp_cmd"]

File: typescript/packages/core/src/commands/builtin/generic/cmp.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { IOResult, materialize, type ByteSource } from '../../../io/types.ts'
import type { PathSpec } from '../../../types.ts'
import type { CommandFnResult, CommandOpts } from '../../config.ts'

const ENC = new TextEncoder()

function octal(n: number): string {
  return '0o' + n.toString(8)
}

function arraysEqual(a: Uint8Array, b: Uint8Array): boolean {
  if (a.byteLength !== b.byteLength) return false
  for (let i = 0; i < a.byteLength; i++) if (a[i] !== b[i]) return false
  return true
}

export async function cmpGeneric(
  paths: PathSpec[],
  opts: CommandOpts,
  stream: (p: PathSpec) => AsyncIterable<Uint8Array>,
): Promise<CommandFnResult> {
  if (paths.length < 2) {
    return [null, new IOResult({ exitCode: 2, stderr: ENC.encode('cmp: requires two paths\n') })]
  }
  const p0 = paths[0]
  const p1 = paths[1]
  if (p0 === undefined || p1 === undefined) return [null, new IOResult()]
  let data1 = await materialize(stream(p0))
  let data2 = await materialize(stream(p1))
  if (typeof opts.flags.i === 'string') {
    const skip = Number.parseInt(opts.flags.i, 10)
    data1 = data1.slice(skip)
    data2 = data2.slice(skip)
  }
  if (typeof opts.flags.n === 'string') {
    const limit = Number.parseInt(opts.flags.n, 10)
    data1 = data1.slice(0, limit)
    data2 = data2.slice(0, limit)
  }
  if (arraysEqual(data1, data2)) return [null, new IOResult()]
  if (opts.flags.s === true) return [null, new IOResult({ exitCode: 1 })]
  if (opts.flags.args_l === true) {
    const outLines: string[] = []
    const limit = Math.min(data1.byteLength, data2.byteLength)
    for (let idx = 0; idx < limit; idx++) {
      if (data1[idx] !== data2[idx]) {
        outLines.push(`${String(idx + 1)} ${octal(data1[idx] ?? 0)} ${octal(data2[idx] ?? 0)}`)
      }
    }
    const out: ByteSource = ENC.encode(outLines.join('\n'))
    return [out, new IOResult({ exitCode: 1 })]
  }
  const limit = Math.min(data1.byteLength, data2.byteLength)
  for (let idx = 0; idx < limit; idx++) {
    if (data1[idx] !== data2[idx]) {
      let line = 1
      for (let k = 0; k < idx; k++) if (data1[k] === 0x0a) line += 1
      let msg = `${p0.original} ${p1.original} differ: char ${String(idx + 1)}, line ${String(line)}`
      if (opts.flags.b === true) {
        msg += ` is ${octal(data1[idx] ?? 0)} ${String.fromCharCode(data1[idx] ?? 0)} ${octal(data2[idx] ?? 0)} ${String.fromCharCode(data2[idx] ?? 0)}`
      }
      return [ENC.encode(msg), new IOResult({ exitCode: 1 })]
    }
  }
  const shorter = data1.byteLength < data2.byteLength ? p0 : p1
  return [ENC.encode(`cmp: EOF on ${shorter.original}`), new IOResult({ exitCode: 1 })]
}
